reporter/src/main/kotlin/reporters/AbstractNoticeReporter.kt

```diff
@@ -0,0 +1,94 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.reporter.reporters
+
+import ch.frankel.slf4k.info
+
+import com.here.ort.model.OrtResult
+import com.here.ort.reporter.Reporter
+import com.here.ort.reporter.ResolutionProvider
+import com.here.ort.utils.log
+import com.here.ort.utils.spdx.getLicenseText
+
+import java.io.File
+import java.util.SortedMap
+import java.util.SortedSet
+
+abstract class AbstractNoticeReporter : Reporter() {
+    override fun generateReport(ortResult: OrtResult, resolutionProvider: ResolutionProvider, outputDir: File): File? {
+        require(ortResult.scanner != null) {
+            "The provided ORT result file does not contain a scan result."
+        }
+
+        val outputFile = File(outputDir, noticeFileName)
+
+        val licenseFindings = getLicenseFindings(ortResult)
+
+        val findingsIterator = licenseFindings.filterNot { (license, _) ->
+            // For now, just skip license references for which SPDX has no license text.
+            license.startsWith("LicenseRef-")
+        }.iterator()
+
+        if (!findingsIterator.hasNext()) {
+            log.info { "Not writing a $noticeFileName file as it would be empty." }
+            return null
+        } else {
+            log.info { "Writing $noticeFileName file to '${outputFile.absolutePath}'." }
+        }
+
+        // Note: Do not use appendln() here as that would write out platform-native line endings, but we want to
+        // normalize on Unix-style line endings for consistency.
+        while (findingsIterator.hasNext()) {
+            val (license, copyrights) = findingsIterator.next()
+
+            var noticeBuilder = StringBuilder()
+
+            copyrights.forEach { copyright ->
+                noticeBuilder.append("$copyright\n")
+            }
+
+            if (copyrights.isNotEmpty()) noticeBuilder.append("\n")
+
+            noticeBuilder.append("${getLicenseText(license, true)}\n")
+
+            // Trim lines and remove consecutive blank lines as the license text formatting in SPDX JSON files is
+            // broken, see https://github.com/spdx/LicenseListPublisher/issues/30.
+            var previousLine = ""
+            val trimmedNoticeLines = noticeBuilder.lines().mapNotNull { line ->
+                val trimmedLine = line.trim()
+                trimmedLine.takeIf { it.isNotBlank() || previousLine.isNotBlank() }
+                        .also { previousLine = trimmedLine }
+            }
+
+            noticeBuilder = StringBuilder(trimmedNoticeLines.joinToString("\n"))
+
+            // Separate notice rows.
+            if (findingsIterator.hasNext()) noticeBuilder.append("\n----\n\n")
+
+            outputFile.appendText(noticeBuilder.toString())
```

Comment:
The full notices text is still small enough to hold in in main memory completely. Thus there is no need to mess around with files for this method - it actually violates single responsitibilty principle. How about extracting the whole while loop into a function, e.g.:
fun createNotices( findings: SortedMap<String, SortedSet>) : String

Reply:
Let's address this separately as part of issue #755 as this is just moved code here.